src/core/Akka.Remote/Transport/DotNetty/BatchWriter.cs

```diff
@@ -0,0 +1,104 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="BatchWriter.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2019 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2019 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Threading.Tasks;
+using DotNetty.Buffers;
+using DotNetty.Common.Concurrency;
+using DotNetty.Transport.Channels;
+
+namespace Akka.Remote.Transport.DotNetty
+{
+    /// <summary>
+    /// INTERNAL API.
+    /// 
+    /// Responsible for batching socket writes together into fewer sys calls to the socket.
+    /// </summary>
+    internal class BatchWriter : ChannelHandlerAdapter
+    {
+        private readonly int _maxPendingWrites;
+        private readonly int _maxPendingMillis;
+        private readonly int _maxPendingBytes;
+
+        public BatchWriter(int maxPendingWrites = 20, int maxPendingMillis = 40, int maxPendingBytes = 128000)
+        {
+            _maxPendingWrites = maxPendingWrites;
+            _maxPendingMillis = maxPendingMillis;
+            _maxPendingBytes = maxPendingBytes;
+        }
+
+        private int _currentPendingWrites = 0;
+        private long _currentPendingBytes;
+
+        public bool HasPendingWrites => _currentPendingWrites > 0;
+
+        public override void HandlerAdded(IChannelHandlerContext context)
+        {
+            ScheduleFlush(context);
+            base.HandlerAdded(context);
+        }
+
+        public override Task WriteAsync(IChannelHandlerContext context, object message)
+        {
+            var write = base.WriteAsync(context, message);
```

Comment:
Might want to add a comment here - need to complete the prior `WriteAsync` call first before we call flush, otherwise the message currently being written may not be included in the flush even though it was counted against `_currentPendingWrites'